File: mobile/reson_app/lib/update/update_checker.dart
import 'dart:async';
import 'dart:convert';

import 'package:http/http.dart' as http;

import 'version.dart';

/// Result of a successful update check: a newer release is available.
class UpdateInfo {
  const UpdateInfo({
    required this.version,
    required this.apkUrl,
    required this.notes,
  });

  /// Latest version, e.g. "1.0.1" (the leading "v" from the tag is stripped).
  final String version;

  /// Direct download URL of the `Reson-Android.apk` release asset.
  final String apkUrl;

  /// Release notes (GitHub release `body`), possibly empty.
  final String notes;
}

/// Checks GitHub Releases for a newer Reson APK. Never throws — any failure
/// (offline, 404 when no releases exist yet, malformed JSON, missing asset)
/// resolves to `null` so callers can fire-and-forget.
class UpdateChecker {
  UpdateChecker({http.Client? client, String? currentVersion})
      : _client = client ?? http.Client(),
        _currentVersionOverride = currentVersion;

  static const _latestReleaseUrl =
      'https://api.github.com/repos/GabrielSoarde/reson/releases/latest';
  static const _apkAssetName = 'Reson-Android.apk';
  static const _timeout = Duration(seconds: 12);

  final http.Client _client;

  /// Lets tests inject the "current" version; otherwise read from package info.
  final String? _currentVersionOverride;

  Future<UpdateInfo?> check() async {
    try {
      final current = _currentVersionOverride ?? await currentAppVersion();
      final r = await _client
          .get(
            Uri.parse(_latestReleaseUrl),
            headers: const {'Accept': 'application/vnd.github+json'},
          )
          .timeout(_timeout);

      // 404 == repo has no published releases yet → silently no update.
      if (r.statusCode != 200) return null;

      final json = jsonDecode(r.body);
      if (json is! Map<String, dynamic>) return null;

      final tag = (json['tag_name'] as String?)?.trim();
      if (tag == null || tag.isEmpty) return null;
      final latest = _stripV(tag);

      // Only offer an update when the release is strictly newer.
      if (compareSemver(latest, current) <= 0) return null;

      final apkUrl = _findApkUrl(json['assets']);
      if (apkUrl == null) return null;

      final notes = (json['body'] as String?)?.trim() ?? '';
      return UpdateInfo(version: latest, apkUrl: apkUrl, notes: notes);
    } catch (_) {
      // Network down, JSON parse failure, etc. — treat as "no update".
      return null;
    }
  }

  void close() => _client.close();

  static String? _findApkUrl(Object? assets) {
    if (assets is! List) return null;
    for (final a in assets) {
      if (a is Map<String, dynamic> &&
          a['name'] == _apkAssetName &&
          a['browser_download_url'] is String) {
        return a['browser_download_url'] as String;
      }
    }
    return null;
  }
}

/// Strips a single leading "v"/"V" from a tag ("v1.0.1" → "1.0.1").
String _stripV(String tag) {
  if (tag.isNotEmpty && (tag[0] == 'v' || tag[0] == 'V')) {
    return tag.substring(1);
  }
  return tag;
}

/// Compares two dotted version strings numerically.
///
/// Tolerant of input quirks: any `+build` suffix is dropped, a leading "v"
/// is ignored, missing components are treated as 0 ("1.0" == "1.0.0"), and
/// non-numeric segments fall back to 0. Returns <0 / 0 / >0 like [Comparable].
int compareSemver(String a, String b) {
  final pa = _parse(a);
  final pb = _parse(b);
  final len = pa.length > pb.length ? pa.length : pb.length;
  for (var i = 0; i < len; i++) {
    final x = i < pa.length ? pa[i] : 0;
    final y = i < pb.length ? pb[i] : 0;
    if (x != y) return x < y ? -1 : 1;
  }
  return 0;
}

List<int> _parse(String v) {
  var s = _stripV(v.trim());
  // Drop the pubspec "+build" suffix and any "-prerelease" tail before splitting.
  final plus = s.indexOf('+');
  if (plus != -1) s = s.substring(0, plus);
  final dash = s.indexOf('-');
  if (dash != -1) s = s.substring(0, dash);
  return s
      .split('.')
      .map((p) => int.tryParse(p.trim()) ?? 0)
      .toList(growable: false);
}

File: src/Soundpad/Update/UpdateChecker.cs
using System.Net;
using System.Text.Json;

namespace Soundpad.Update;

/// <summary>
/// Metadata describing a newer release found on GitHub.
/// </summary>
public sealed record UpdateInfo(string Version, string DownloadUrl, string ReleaseNotes);

/// <summary>
/// Lightweight auto-updater that polls the public GitHub Releases API and, if a
/// newer <c>ResonSetup.exe</c> is available, downloads it and re-runs it. The
/// installer's in-place upgrade (same Inno Setup AppId) handles replacing the
/// files and preserving <c>%LOCALAPPDATA%\Reson\</c> data.
///
/// <para>Network failures are intentionally swallowed: offline / GitHub down /
/// no releases published all collapse to "no update available" (null) so the
/// check can never crash or block app startup.</para>
/// </summary>
public sealed class UpdateChecker
{
    // GitHub's API requires a User-Agent header (403 without one) and the
    // /releases/latest endpoint returns 404 when there are zero published
    // (non-prerelease) releases.
    public const string ReleasesLatestUrl = "https://api.github.com/repos/GabrielSoarde/reson/releases/latest";
    public const string UserAgent = "Reson-Updater";
    public const string SetupAssetName = "ResonSetup.exe";

    private readonly HttpClient _http;
    private readonly string _currentVersion;

    public UpdateChecker(HttpClient? http = null, string? currentVersion = null)
    {
        _http = http ?? new HttpClient { Timeout = TimeSpan.FromSeconds(15) };
        _currentVersion = currentVersion ?? AppVersion.Current;
    }

    /// <summary>
    /// Determines whether <paramref name="latestTag"/> (e.g. "v1.0.1") is a newer
    /// version than <paramref name="currentVersion"/> (e.g. "1.0.0"). Strips a
    /// single leading "v" from each side and compares with <see cref="Version"/>
    /// semantics. Returns false if either side fails to parse (malformed → treat
    /// as "no update"), or if they're equal / older.
    /// </summary>
    public static bool IsNewer(string? latestTag, string? currentVersion)
    {
        if (!Version.TryParse(StripLeadingV(latestTag), out var latest)) return false;
        if (!Version.TryParse(StripLeadingV(currentVersion), out var current)) return false;
        return latest > current;
    }

    private static string? StripLeadingV(string? tag)
    {
        if (string.IsNullOrWhiteSpace(tag)) return null;
        tag = tag.Trim();
        return (tag.StartsWith("v", StringComparison.OrdinalIgnoreCase) ? tag[1..] : tag).Trim();
    }

    /// <summary>
    /// Queries GitHub for the latest release. Returns an <see cref="UpdateInfo"/>
    /// when a strictly-newer release with a <c>ResonSetup.exe</c> asset exists,
    /// otherwise null. Never throws — all network/parse errors and a 404 (no
    /// releases yet) map to null.
    /// </summary>
    public async Task<UpdateInfo?> CheckAsync(CancellationToken ct = default)
    {
        try
        {
            using var req = new HttpRequestMessage(HttpMethod.Get, ReleasesLatestUrl);
            req.Headers.UserAgent.ParseAdd(UserAgent);
            req.Headers.Accept.ParseAdd("application/vnd.github+json");

            using var resp = await _http.SendAsync(req, HttpCompletionOption.ResponseHeadersRead, ct).ConfigureAwait(false);

            // 404 = no published (non-prerelease) releases yet → not an error.
            if (resp.StatusCode == HttpStatusCode.NotFound) return null;
            if (!resp.IsSuccessStatusCode) return null;

            await using var stream = await resp.Content.ReadAsStreamAsync(ct).ConfigureAwait(false);
            using var doc = await JsonDocument.ParseAsync(stream, cancellationToken: ct).ConfigureAwait(false);
            var root = doc.RootElement;

            if (!root.TryGetProperty("tag_name", out var tagEl)) return null;
            var tag = tagEl.GetString();
            if (string.IsNullOrWhiteSpace(tag)) return null;

            if (!IsNewer(tag, _currentVersion)) return null;

            // Find the ResonSetup.exe asset's download URL.
            string? downloadUrl = null;
            if (root.TryGetProperty("assets", out var assets) && assets.ValueKind == JsonValueKind.Array)
            {
                foreach (var asset in assets.EnumerateArray())
                {
                    if (asset.TryGetProperty("name", out var nameEl) &&
                        string.Equals(nameEl.GetString(), SetupAssetName, StringComparison.OrdinalIgnoreCase) &&
                        asset.TryGetProperty("browser_download_url", out var urlEl))
                    {
                        downloadUrl = urlEl.GetString();
                        break;
                    }
                }
            }

            // A newer tag without a usable installer asset is useless to us.
            if (string.IsNullOrWhiteSpace(downloadUrl)) return null;

            var notes = root.TryGetProperty("body", out var bodyEl) ? bodyEl.GetString() ?? "" : "";
            var version = StripLeadingV(tag) ?? tag;

            return new UpdateInfo(version, downloadUrl, notes);
        }
        catch
        {
            // Offline, DNS failure, timeout, malformed JSON, cancellation — all
            // collapse to "no update". The updater must never crash the app.
            return null;
        }
    }

    /// <summary>
    /// Downloads the installer to <c>%TEMP%\ResonSetup-&lt;version&gt;.exe</c> and
    /// returns its path. Reports 0..1 progress when the server sends a
    /// Content-Length. Throws on failure (the caller surfaces the error).
    /// </summary>
    public async Task<string> DownloadAsync(UpdateInfo info, IProgress<double>? progress = null, CancellationToken ct = default)
    {
        var dest = Path.Combine(Path.GetTempPath(), $"ResonSetup-{info.Version}.exe");

        using var req = new HttpRequestMessage(HttpMethod.Get, info.DownloadUrl);
        req.Headers.UserAgent.ParseAdd(UserAgent);

        using var resp = await _http.SendAsync(req, HttpCompletionOption.ResponseHeadersRead, ct).ConfigureAwait(false);
        resp.EnsureSuccessStatusCode();

        var total = resp.Content.Headers.ContentLength;
        await using var src = await resp.Content.ReadAsStreamAsync(ct).ConfigureAwait(false);
        await using var dst = new FileStream(dest, FileMode.Create, FileAccess.Write, FileShare.None);

        var buffer = new byte[81920];
        long read = 0;
        int n;
        while ((n = await src.ReadAsync(buffer, ct).ConfigureAwait(false)) > 0)
        {
            await dst.WriteAsync(buffer.AsMemory(0, n), ct).ConfigureAwait(false);
            read += n;
            if (total is > 0)
                progress?.Report((double)read / total.Value);
        }
        progress?.Report(1.0);
        return dest;
    }

    /// <summary>
    /// Launches the downloaded installer and shuts the app down so the in-place
    /// upgrade can replace the running exe. Best-effort: if the launch fails the
    /// app keeps running (caller can surface the error).
    /// </summary>
    public void RunInstallerAndExit(string installerPath)
    {
        System.Diagnostics.Process.Start(new System.Diagnostics.ProcessStartInfo
        {
            FileName = installerPath,
            UseShellExecute = true,
        });
        // The installer needs the running Reson.exe gone to overwrite files.
        // Environment.Exit tears down both the WPF UI thread and the Kestrel
        // app.Run() loop on the main thread cleanly enough for the upgrade.
        Environment.Exit(0);
    }
}
